components/renku_data_services/data_connectors/db.py

```diff
@@ -477,11 +477,47 @@ def __init__(
         data_connector_repo: DataConnectorRepository,
         user_repo: UserRepo,
         secret_service_public_key: rsa.RSAPublicKey,
+        authz: Authz,
     ) -> None:
         self.session_maker = session_maker
         self.data_connector_repo = data_connector_repo
         self.user_repo = user_repo
         self.secret_service_public_key = secret_service_public_key
+        self.authz = authz
+
+    async def get_data_connectors_with_secrets(
+        self,
+        user: base_models.APIUser,
+        project_id: ULID,
+    ) -> AsyncIterator[models.DataConnectorWithSecrets]:
+        """Get all data connectors and their secrets for a project."""
+        if user.id is None:
+            raise errors.UnauthorizedError(message="You do not have the required permissions for this operation.")
+
+        can_read_project = await self.authz.has_permission(user, ResourceType.project, project_id, Scope.READ)
+        if not can_read_project:
+            raise errors.MissingResourceError(
+                message=f"The project ID with {project_id} does not exist or you dont have permission to access it"
+            )
+
+        async with self.session_maker() as session:
+            stmt = select(schemas.DataConnectorORM).where(
+                schemas.DataConnectorORM.project_links.any(
+                    schemas.DataConnectorToProjectLinkORM.project_id == project_id
+                ),
+                or_(
+                    # Data connectors with secrets for the specific user
+                    schemas.DataConnectorORM.secrets.any(
+                        schemas.DataConnectorSecretORM.user_id == user.id,
+                    ),
+                    # Data connectors without any secrets
+                    # See: https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#exists-forms-has-any
+                    ~schemas.DataConnectorORM.secrets.any(),
```

Comment:
I'm pretty confused by that part of the statement. Why include it at all? If you do the following:
```python
stmt = select(schemas.DataConnectorORM).where(
                schemas.DataConnectorORM.project_links.any(
                    schemas.DataConnectorToProjectLinkORM.project_id == project_id
                )
            )
```

Would that not select all data connectors linked to a project?

Reply:
Yes you are right. I will change it and then "manually" filter the secrets.

Reply:
You can just use the `get_data_connector_secrets()` method below.